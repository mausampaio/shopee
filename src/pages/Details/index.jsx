import React, { useContext, useEffect, useState } from "react";
import { useHistory, useParams } from "react-router-dom";
import { ButtonWrapper, ButtonContainer } from "./styles";
import { Container } from "../../components/Container";
import { ProductCardDetails } from "../../components/ProductCard";
import { Store } from "../../context";
import { Loading } from "../../components/Loading";
import SearchForm from "../../components/SearchForm";
import useWindowDimensions from "../../hooks/useWindowDimensions";
import { Button } from "../../components/Button";

function Details(props) {
  const context = useContext(Store);
  const { id } = useParams();
  const { width } = useWindowDimensions();
  const [product, setProduct] = useState(null);

  useEffect(() => {
    if (context.products.length !== 0) {
      const selectedProd = context.products.find((prod) => prod.id === +id);
      setProduct(selectedProd);
    }
  }, [context.products, id]);

  const history = useHistory();

  const handleGoBack = () => {
    history.goBack();
  };

  return (
    <Container>
      {context.loading ? (
        <Loading />
      ) : (
        <>
          <h1>{product?.name || "Produto n√£o encontrado!"}</h1>
          {width <= 575 ? <SearchForm /> : null}
          <div>{product && <ProductCardDetails product={product} />}</div>
          <ButtonWrapper>
            <ButtonContainer>
              <Button onClick={handleGoBack}>Voltar</Button>
            </ButtonContainer>
          </ButtonWrapper>
        </>
      )}
    </Container>
  );
}

export default Details;
