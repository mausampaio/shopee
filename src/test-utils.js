import { render } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";

export const mockProduct = {
  id: 1,
  name: "Produto de teste",
  price: 1.0,
  description: "Produto gerado para testes",
  url: "produto-de-teste",
  image: "./produto-teste.jpg",
};

export const mockCart = {
  id: mockProduct.id,
  name: mockProduct.name,
  price: mockProduct.price,
  quantity: 10,
};

export const mockContext = {
  cart: [],
  products: [],
  product: {},
  loading: false,
  getCartTotal: () => 0,
};

const RouterProvider = ({ children }) => {
  return <BrowserRouter>{children}</BrowserRouter>;
};

const customRender = (ui, options) => {
  return render(<RouterProvider>{ui}</RouterProvider>, { ...options });
};

export * from "@testing-library/react";

export { customRender as render };
