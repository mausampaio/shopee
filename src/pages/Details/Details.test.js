import Details from ".";
import { Store } from "../../context";
import { mockContext, mockProduct, render } from "../../test-utils";

// mock a param with value 'id = 1'
jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useParams: () => ({ id: 1 }),
}));

const ContextProvider = ({ children }) => (
  <Store.Provider value={{ ...mockContext, products: [mockProduct] }}>
    {children}
  </Store.Provider>
);

test("should render a product detail using id=1 on params", () => {
  const { getByText } = render(<Details />, {
    wrapper: ContextProvider,
  });
  expect(getByText(mockProduct.name)).toBeInTheDocument();
});
