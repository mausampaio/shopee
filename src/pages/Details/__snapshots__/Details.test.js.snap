// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<DocumentFragment>
  .c8 {
  padding-top: 32px;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row-reverse;
  -ms-flex-direction: row-reverse;
  flex-direction: row-reverse;
  margin-bottom: 32px;
}

.c9 {
  width: 226px;
}

.c0 {
  width: 80%;
  max-width: 1280px;
  margin: 0 auto;
}

.c7 {
  width: 100%;
  min-width: 44px;
  height: 44px;
  background-color: var(--secondary);
  border-radius: 4px;
  color: var(--primary);
  font-size: 18px;
  font-weight: 500;
  cursor: pointer;
}

.c6 {
  width: 226px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  background-color: var(--primary);
  width: 258px;
  padding: 16px;
  border-radius: 7px;
  box-shadow: 0 4px 8px rgba(0,0,0,0.1);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  padding: 32px;
}

.c2 {
  width: 226px;
  height: 226px;
  overflow: hidden;
  background-color: var(--primary);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  margin-right: 32px;
}

.c2 img {
  max-width: 100%;
  max-height: 100%;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-size: 32px;
  line-height: 24px;
  font-weight: 500;
  margin-bottom: 8px;
  display: block;
}

.c4 span {
  font-size: 14px;
}

.c5 {
  font-size: 12px;
  line-height: 24px;
  font-weight: 400;
  margin-bottom: 16px;
  display: block;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c3 p {
  margin-top: 8px;
}

@media screen and (max-width:575px) {
  .c8 {
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
    -webkit-box-pack: center;
    -webkit-justify-content: center;
    -ms-flex-pack: center;
    justify-content: center;
  }
}

@media screen and (max-width:768px) {
  .c0 {
    width: 90%;
  }
}

@media screen and (max-width:575px) {
  .c1 {
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
    -webkit-align-items: center;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
  }
}

@media screen and (max-width:575px) {
  .c2 {
    margin: 0;
  }
}

@media screen and (max-width:575px) {
  .c3 {
    -webkit-align-items: center;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
  }

  .c3 p {
    margin: 16px 0;
  }
}

<div
    class="c0"
  >
    <h1>
      Produto de teste
    </h1>
    <div>
      <div
        class="c1"
      >
        <div
          class="c2"
        >
          <img
            alt="Produto de teste"
            src="./produto-teste.jpg"
          />
        </div>
        <div
          class="c3"
        >
          <p>
            Produto gerado para testes
          </p>
          <div
            style="width: 100%;"
          >
            <div
              class="c4"
            >
              <span
                style="font-size: 20px; line-height: 24px; font-weight: 500; padding-top: 0px;"
              >
                R$
              </span>
              <span
                style="font-size: 32px; line-height: 24px; font-weight: 500;"
              >
                 1
              </span>
              <span
                style="font-size: 16px; line-height: 24px; font-weight: 500;"
              >
                ,00
              </span>
            </div>
            <span
              class="c5"
            >
              ou 10x de R$ 0,10
            </span>
          </div>
          <div
            class="c6"
          >
            <button
              class="c7"
            >
              Comprar
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="c8"
    >
      <div
        class="c9"
      >
        <button
          class="c7"
        >
          Voltar
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
